code/modules/mob/camera/imaginary_friend.dm

```diff
@@ -0,0 +1,282 @@
+	if(!message)
+		return
+
+	if(client)
+		if(client.prefs.muted & MUTE_IC)
+			to_chat(src, "You cannot send IC messages (muted).")
+			return
+
+		if(client.handle_spam_prevention(message, MUTE_IC))
+			return
+
+	friend_talk(message)
+
+/// shows langchat and speech text to the owner and friend, and sends speech text to dchat
+/mob/camera/imaginary_friend/proc/friend_talk(message)
+	message = capitalize(trim(copytext_char(sanitize(message), 1, MAX_MESSAGE_LEN)))
+
+	if(!message)
+		return
+
+	var/rendered = "<span class='game say'><span class='name'>[name]</span> <span class='message'>[say_quote(message)] \"[message]\"</span></span>"
+	var/dead_rendered = "<span class='game say'><span class='name'>[name] (imaginary friend of [owner])</span> <span class='message'>[say_quote(message)] \"[message]\"</span></span>"
+
+	to_chat(owner, "[rendered]")
+	to_chat(src, "[rendered]")
+	if(!hidden)
+		var/list/send_to = list()
+		if(!owner.client?.prefs.lang_chat_disabled)
+			send_to += owner
+		if(!client?.prefs.lang_chat_disabled)
+			send_to += src
+		if(length(send_to))
+			langchat_speech(message, send_to, GLOB.all_languages, skip_language_check = TRUE)
+
+	//speech bubble
+	var/mutable_appearance/MA = mutable_appearance('icons/mob/effects/talk.dmi', src, "default[say_test(message)]", FLY_LAYER)
+	MA.appearance_flags = APPEARANCE_UI_IGNORE_ALPHA
+	INVOKE_ASYNC(GLOBAL_PROC, GLOBAL_PROC_REF(flick_overlay_to_clients), MA, owner.client ? list(client, owner.client) : list(client), 3 SECONDS)
+
+	for(var/mob/ghost as anything in GLOB.dead_mob_list)
+		if(isnewplayer(ghost) || src == ghost)
+			continue
+		var/link = "<a href='byond://?src=\ref[ghost];track=\ref[src]'>F</a>"
+		to_chat(ghost, "[dead_rendered] ([link])")
+
+/mob/camera/imaginary_friend/Move(newloc, Dir = 0)
+	if(world.time < move_delay)
+		return FALSE
+
+	if(get_dist(src, owner) > 9)
+		recall()
+		move_delay = world.time + 10
+		return FALSE
+
+	forceMove(newloc)
+	move_delay = world.time + 1
+
+/mob/camera/imaginary_friend/forceMove(atom/destination)
+	dir = get_dir(get_turf(src), destination)
+	loc = destination
+	show()
+
+/// returns the friend to the owner
+/mob/camera/imaginary_friend/proc/recall()
+	if(QDELETED(owner))
+		deactivate()
+		return FALSE
+	if(loc == owner)
+		return FALSE
+	forceMove(owner)
+
+/// logs the imaginary friend's removal, ghosts them and cleans up the friend
+/mob/camera/imaginary_friend/proc/deactivate()
+	log_admin("[key_name(src)] stopped being imaginary friend of [key_name(owner)].")
+	message_admins("[key_name_admin(src)] stopped being imaginary friend of [key_name_admin(owner)].")
+	ghostize(TRUE, TRUE)
+	qdel(src)
+
+/mob/camera/imaginary_friend/ghostize(can_reenter_corpse = FALSE, aghosted = FALSE)
+	if(QDELING(src))
+		return
+
+	icon = friend_image
+	mouse_opacity = MOUSE_OPACITY_ICON
+	var/mob/ghost = ..()
+	ghost.mind.original = aghosted_original_mob
```

Comment:
Possible mind check is required here, sometimes it doesn't seem to exist and I'm not certain why yet